### PR TITLE
fix: Markdown hover handling causing DevCard errors

### PR DESCRIPTION
**Pull request by ilasw**
## Changes

Before: we were using ref to get the markdown div and put event listeners for every mention link.
Now: I've put a single event listener `onMouseOverCapture` that checks the current hovered element and if is a link element set userId.

I've also removed the _now-no-more-needed_ ref in div. It was in conflict with `React.cloneElement` inside `SimpleTooltip` component.

Pasted Graphic

## Events

Did you introduce any new tracking events?

## Experiment

Did you introduce any new experiments?

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

### Preview domain
Actual Prod: https://app.daily.dev/posts/we-made-daily-dev-faster-see-how-we-did-it-uzorvcrrs
Demo, now mention is working: https://fix-mention-hover-error.preview.app.daily.dev/posts/we-made-daily-dev-faster-see-how-we-did-it-uzorvcrrs

### CODE REVIEW

File: packages/shared/src/components/Markdown.tsx [hunk 1]
```diff
@@ -101,17 +92,16 @@ export default function Markdown({
         offset,
         visible: !!userId,
         onShow: cancelUserClearing,
+        onHide: clearUser,
         appendTo: appendTooltipTo || globalThis?.document?.body || 'parent',
       }}
-      onMouseEnter={cancelUserClearing}
-      onMouseLeave={clearUser}
```
Comment (ilasw) [inline packages/shared/src/components/Markdown.tsx:107]: Those were passed but never used inside, just removed for clarity

Comment (rebelchris) [reply packages/shared/src/components/Markdown.tsx:107]: I think they where needed for the clearing right?
Now we can hover multiple users and it stays:
Screenshot 2024-11-06 at 08 11 05

File: packages/shared/src/components/Markdown.tsx [hunk 2]
```diff
@@ -35,63 +47,42 @@ export default function Markdown({
   );
 
   useEffect(() => {
-    if (!content || !ref?.current) {
+    if (data || !userId) {
       return;
     }
+    fetchInfo();
+    // @NOTE see https://dailydotdev.atlassian.net/l/cp/dK9h1zoM
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (rebelchris) [inline packages/shared/src/components/Markdown.tsx:55]: Probably remove this and just add all deps, they should be safe

Comment (rebelchris) [reply packages/shared/src/components/Markdown.tsx:55]: We can actually probably just do a cleanup effect here that clears user?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Just some minor issues on testing.
Overall it looks good though.

File: packages/shared/src/components/Markdown.tsx (hunk 1 above)
Comment (ilasw) [reply packages/shared/src/components/Markdown.tsx:107]: Nice catch, thanks for reporting! Seems that the DevCard is not closing if the mention link is the last element hovered inside the Markdown wrapper.  I'm adding a clearUser onMouseLeave in order to fix it.

File: packages/shared/src/hooks/useProfileTooltip.ts [hunk 3]
```diff
@@ -57,12 +57,9 @@ export const useProfileTooltip = ({
...
 
-  return useMemo(
-    () => ({
-      data,
-      isLoading: isPending,
-      fetchInfo: () => setShouldFetch(true),
-    }),
-    [data, isPending, setShouldFetch],
-  );
+  return {
```
Comment (rebelchris) [inline packages/shared/src/hooks/useProfileTooltip.ts:60]: Shouldn't we need the memo here for data changes still?

**Review by rebelchris — APPROVED** [review]
Nice works like a charm now, just not sure if there's now an edge-case with the memo removal maybe?

File: packages/shared/src/hooks/useProfileTooltip.ts (hunk 3 above)
Comment (capJavert) [reply packages/shared/src/hooks/useProfileTooltip.ts:60]: not really, you never use [the whole object](https://dailydotdev.atlassian.net/wiki/spaces/HAN/pages/248020993/Guidelines#Other-custom-hooks:~:text=We%20do%20not%20memoize%20return%20object%20because%20usage%20of%20our%20business%20logic%20hooks%20should%20be%20through%20destructuring%20assignment.) returned from hook

Comment (ilasw) [reply packages/shared/src/hooks/useProfileTooltip.ts:60]: I've checked the usage, and useProfileTooltip is only used here, so this change is quite safe, imo.
`data` and `isPending` are handled/cached by react-query, we should care only about `setShouldFetch`, which shouldn't cause issues 👀

**Pull request merged**